doc/manual/contract.md

```diff
@@ -0,0 +1,175 @@
+# Contract Programming
+
+Nit supports contract programming like Eiffel. Contracts can be seen as the materialization of specifications in the code itself. The idea is to define the behavior of a property using conditions which will be evaluated on runtime.
+
+In Nit all the conditions of contract must be expressed using boolean expression in the same way as an assertion. Note that the potential routines that compose the condition of the contract should have no side effects.
+
+## Method/Attribute contract
+
+Methods and attributes (setter) can be associated with two types of contracts, entry contracts (expect) and exit contracts (ensure). Together, they make it possible to define the rules to be respected between the called and the caller. If the caller fulfills the `expect` condition on entry then the caller can assume that the called fulfills the `ensure` condition on exit.
+
+Both are declared, between the definition of the signature followed by the keyword `is` and the block start keyword `do`.
+
+### Expect
+
+We define a precondition using the annotation `expect`. Here is an example with a `pop` method which removes the last element of the stack. It defines as a precondition to its call that the stack must not be empty.
+
+~~~
+	fun pop
+	is
+		expect(not_empty)
+	do
+		# Remove the last item on the stack
+	end
+~~~
+
+### Ensure
+
+We define a postcondition using the annotation `ensure`. Here is an example with a `push` method which adds the element received as an argument (`object`) at the top of the stack. It defines as guaranteed to the caller that the object will be added at the top of the stack `self [size] == object`.
+
+~~~
+	fun push(object: T)
+	is
+		ensure(self[size] == object)
+	do
+		# Add the given argument `object` on top of the stack
+	end
+~~~
+
+The `ensure` contract offers the possibility of referring to the return value of the method, using the keyword `result`. Example the `get_at` method which returns the object which is stored at the index received as an argument (`index`). It guarantees the caller that the returned element represented by `result` corresponds to the element stored in the stack with the given index `result == self [index]`.
+
+~~~
+	fun get_at(index: Int): T
+	is
+		ensure(result == self[index])
+	do
+		# Return the element at the given index.
+	end
+~~~
+
+The `ensure` contract allows to refer to the values in a context before the execution of the method. For this we use the keyword `old` followed by the expression to be evaluated in parenthesis. Example with the `pop` method introduced previously, the contract guarantees that, when it's called to remove the last element from the stack the total size at exit must be equal to the previous size minus one `old (size) - 1 = = size`.
```

Comment:
"The `ensure` contract allows to refer to the values in a context before the execution of the method. For this we use the keyword `old` followed by the expression to be evaluated in parenthesis"
Pour:
Using the keyword `old` we can refer to any arguments and instance variable before the method call.